src/lib/operators/aggregate_hash.cpp

```diff
@@ -215,54 +215,80 @@ void AggregateHash::_aggregate() {
       const auto column_id = _groupby_column_ids.at(group_column_index);
       const auto data_type = input_table->column_data_type(column_id);
 
+      const auto chunk_count = input_table->chunk_count();
+
       resolve_data_type(data_type, [&](auto type) {
         using ColumnDataType = typename decltype(type)::type;
 
-        /*
-        Store unique IDs for equal values in the groupby column (similar to dictionary encoding).
-        The ID 0 is reserved for NULL values. The combined IDs build an AggregateKey for each row.
-        */
-
-        // This time, we have no idea how much space we need, so we take some memory and then rely on the automatic
-        // resizing. The size is quite random, but since single memory allocations do not cost too much, we rather
-        // allocate a bit too much.
-        auto temp_buffer = boost::container::pmr::monotonic_buffer_resource(1'000'000);
-        auto allocator = PolymorphicAllocator<std::pair<const ColumnDataType, AggregateKeyEntry>>{&temp_buffer};
-
-        auto id_map = std::unordered_map<ColumnDataType, AggregateKeyEntry, std::hash<ColumnDataType>, std::equal_to<>,
-                                         decltype(allocator)>(allocator);
-        AggregateKeyEntry id_counter = 1u;
-
-        const auto chunk_count = input_table->chunk_count();
-        for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
-          const auto chunk_in = input_table->get_chunk(chunk_id);
-          if (!chunk_in) continue;
-
-          const auto base_segment = chunk_in->get_segment(column_id);
-
-          ChunkOffset chunk_offset{0};
-          segment_iterate<ColumnDataType>(*base_segment, [&](const auto& position) {
-            if (position.is_null()) {
+        if constexpr (std::is_same_v<ColumnDataType, int32_t>) {
+          // For values smaller than AggregateKeyEntry, we can use the value itself as an AggregateKeyEntry. We cannot
+          // do this for values with the same size as we need have a special NULL value.
+          for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
+            const auto chunk_in = input_table->get_chunk(chunk_id);
+            const auto base_segment = chunk_in->get_segment(column_id);
+            ChunkOffset chunk_offset{0};
+            segment_iterate<ColumnDataType>(*base_segment, [&](const auto& position) {
               if constexpr (std::is_same_v<AggregateKey, AggregateKeyEntry>) {
-                keys_per_chunk[chunk_id][chunk_offset] = 0u;
+                if (position.is_null()) {
+                  keys_per_chunk[chunk_id][chunk_offset] = 0;
+                } else {
+                  keys_per_chunk[chunk_id][chunk_offset] = position.value() + 1;
+                }
               } else {
-                keys_per_chunk[chunk_id][chunk_offset][group_column_index] = 0u;
+                if (position.is_null()) {
+                  keys_per_chunk[chunk_id][chunk_offset][group_column_index] = 0;
```

Comment:
`!position.is_null() * (position.value() + 1)`?

Reply:
Benchmarking

Reply:
```diff
 +----------------+---------------+-------+----------------+-------+------------+---------+
 | Benchmark      | prev. iter/s  | runs  | new iter/s     | runs  | change [%] | p-value |
 +----------------+---------------+-------+----------------+-------+------------+---------+
 | TPC-H 01       | 0.60118240118 | 37    | 0.598081767559 | 36    | -1%        |  0.4720 |
 | TPC-H 02       | 68.4920883179 | 4110  | 68.7720794678  | 4127  | +0%        |  0.5938 |
 | TPC-H 03       | 7.33792161942 | 441   | 7.20852565765  | 433   | -2%        |  0.0018 |
 | TPC-H 04       | 6.91902589798 | 416   | 6.93402576447  | 417   | +0%        |  0.6708 |
 | TPC-H 05       | 3.61497426033 | 217   | 3.72428154945  | 224   | +3%        |  0.0104 |
 | TPC-H 06       | 201.134384155 | 12069 | 201.077758789  | 12065 | -0%        |  0.8469 |
 | TPC-H 07       | 7.51439905167 | 451   | 7.47043275833  | 449   | -1%        |  0.3981 |
 | TPC-H 08       | 6.35709524155 | 382   | 6.37827968597  | 383   | +0%        |  0.5325 |
 | TPC-H 09       | 1.52550828457 | 92    | 1.5229023695   | 92    | -0%        |  0.7316 |
 | TPC-H 10       | 2.65954232216 | 160   | 2.66595959663  | 160   | +0%        |  0.7728 |
 | TPC-H 11       | 30.1973304749 | 1812  | 30.213684082   | 1813  | +0%        |  0.6766 |
 | TPC-H 12       | 8.76660633087 | 526   | 8.73143959045  | 524   | -0%        |  0.0000 |
 | TPC-H 13       | 2.27395629883 | 137   | 2.30064845085  | 139   | +1%        |  0.0000 |
 | TPC-H 14       | 45.9030075073 | 2755  | 46.2060508728  | 2773  | +1%        |  0.0000 |
 | TPC-H 15       | 84.651802063  | 5080  | 85.2034759521  | 5113  | +1%        |  0.0000 |
 | TPC-H 16       | 7.06166124344 | 424   | 7.06917524338  | 425   | +0%        |  0.7488 |
 | TPC-H 17       | 4.69012594223 | 282   | 4.70160293579  | 283   | +0%        |  0.7404 |
 | TPC-H 18       | 1.18020284176 | 71    | 1.20913636684  | 73    | +2%        |  0.0763 |
-| TPC-H 19       | 8.94591522217 | 537   | 8.18227481842  | 492   | -9%        |  0.0000 |
 | TPC-H 20       | 27.0636177063 | 1624  | 26.9943580627  | 1620  | -0%        |  0.3391 |
 | TPC-H 21       | 1.18508327007 | 72    | 1.14148116112  | 69    | -4%        |  0.0023 |
 | TPC-H 22       | 14.209446907  | 853   | 14.2542133331  | 856   | +0%        |  0.3561 |
 | geometric mean |               |       |                |       | -0%        |         |
 +----------------+---------------+-------+----------------+-------+------------+---------+
```

I'd shrug off Q19, but the rest does not seem to make a difference big enough to justify an approach that is arguably less intuitive.

Reply:
Ack.